FIX: Refactor of ``build_path`` and inner machinery

oesteban opened the pull request:
Although some issues have been addressed (#516), it seems that the build_path function is still not digesting correctly some patterns.

This PR:
  - [x] reduces the number of lines to implement the functionality (by removing the ``replace_entities`` function, which is now integrated withing ``build_path``)
  - [x] adds a good bunch of tests (and, in particular, the failing examples of #515) to quickly identify regressions.
  - [x] revises some of the ``default_path_patterns`` in the config.
  - [x] enables doctests (to support the next item)
  - [x] adds the writing module to the docs: https://117-62954915-gh.circle-artifacts.com/0/root/project/doc/_build/html/generated/bids.layout.writing.html#module-bids.layout.writing

oesteban commented on bids/layout/writing.py:
```diff
@@ -87,27 +44,136 @@ def build_path(entities, path_patterns, strict=False):
+        entities['extension'] = entities['extension'][1:]
+
+    # Drop None and empty-strings
+    entities = {k: v for k, v in entities.items() if v}
```
I have to revise this: integer 0 will be dropped. Which brings the question of whether we should enforce strings given that we don't have formatters (e.g.., zero-padded ints should be formatted already)

Booleans, empty strings and Nones should always be dropped, correct?

cc/ @effigies

tyarkoni replied:
Yeah I think we need to keep `0`, as it could be something like `run-0`, unless the spec explicitly states that indexing must be from `1`. Ideally we would do some type checking for known entities and/or let the user pass in type declarations in the patterns, but that seems like more trouble than it's worth. I agree that we should probably just assume that the user is passing in all values exactly as they should appear. So if users pass ints, they're going to be naively converted to strings, and if the user doesn't like that, they can pass strings (we should make sure the docstring makes that clear).

oesteban replied:
Added just 0 as a white-list kind of approach 👍

tyarkoni commented on bids/layout/writing.py:
```diff
@@ -6,57 +6,14 @@
+__all__ = ['build_path', 'write_contents_to_file']
 
-
-def replace_entities(entities, pattern):
```
Are we confident nobody will ever want to use this directly, outside of the context of path-building? I don't remember the original motivation for exposing this at module level, but I suspect that was the consideration....

oesteban replied:
Having this method separated included an extra run of the regex per entity. Now everything is done in the internal loop, and the actual replacement uses python's formatting (which is accessible to anyone). I would see some interest in getting the pattern once interpreted instead and allow the user do the formatting... but honestly I'm not sure it is worth the separation.

tyarkoni replied:
Cool, let's leave it as-is for now then.

tyarkoni commented on bids/layout/writing.py:
```diff
@@ -87,27 +44,136 @@ def build_path(entities, path_patterns, strict=False):
     path_patterns = listify(path_patterns)
 
+    # One less source of confusion
+    ext = entities.get('extension') or ''
```
Keep in mind that until (I think) version 0.12, we're committed to supporting both `extensions` and `extension` in `BIDSLayout.get()`. Not sure if you want to maintain the same approach here.

oesteban replied:
Oh, yes - but the default patterns we already fixed to "extension" so I guess there's no point in adding support for extensions in this particular method.

tyarkoni commented on bids/layout/writing.py:
```diff
@@ -87,27 +44,136 @@ def build_path(entities, path_patterns, strict=False):
 
+    # One less source of confusion
+    ext = entities.get('extension') or ''
+    if ext.startswith('.'):
```
Could just do `entities['extension'] = entities.get('extension', '').lstrip('.')`, I think.

oesteban replied:
Yup, amended (commit will be pushed shortly)

tyarkoni commented on bids/layout/writing.py:
```diff
@@ -87,27 +44,136 @@ def build_path(entities, path_patterns, strict=False):
+        # Expand options within valid values and
+        # check whether entities provided have acceptable value
+        tmp_entities = entities.copy()  # Do not modify the original query
+        for fmt, name, valid, defval in zip(
```
Do we need to `zip` twice? Isn't the result of this 2nd zip just the same as looping over `entities_matched`?

oesteban replied:
thx! not really. Amended.

oesteban commented:
Although I seem to have write permissions to the repo, I'd rather not merge anything myself (esp. if I'm the requester). Please let me know if I did not successfully address all the questions and comments.

tyarkoni commented:
LGTM. Let's leave open for a day or so in case @effigies wants to review, otherwise I'm fine merging.

effigies commented:
Looking over this now. One thing I'm noting is that suffix is now optional for a lot of patterns. I haven't pulled and tested this yet, but I think the following will occur:

```Python
>>> layout.build_path({'subject': '01'})
'sub-01/dwi/sub-01_dwi.nii.gz'

>>> layout.build_path({'subject': '01', 'task': 'rest'})
'sub-01/meg/sub-01_task-rest_coordsystem.json'
```

These seem like unintuitive behaviors. Anyway, want to start discussion on that while I look more closely.

effigies commented:
Ah:

```Python
In [1]: import bids                                                                                 

In [2]: bids.__version__                                                                            
Out[2]: '0.10.0+38.ga222c50'

In [3]: layout = bids.BIDSLayout('.')                                                               

In [4]: layout.build_path({'subject': '01'})                                                        
Out[4]: 'sub-01/dwi/sub-01_dwi.nii.gz'

In [5]: layout.build_path({'subject': '01', 'task': 'rest'})                                        
Out[5]: 'sub-01/dwi/sub-01_dwi.nii.gz'
```

I think it'd be good to add a test to make sure all entities passed are used.

oesteban commented:
I think that what ``strict=True`` is for right?

```
In [6]: build_path({'subject': '01', 'task': 'rest'}, default_path_patterns)
Out[6]: 'sub-01/dwi/sub-01_dwi.nii.gz'

In [7]: build_path({'subject': '01', 'task': 'rest'}, default_path_patterns, strict=True)
Out[7]: 'sub-01/sub-01_task-rest_events.tsv'
```

oesteban commented:
That said, we could remove the default for suffixes, I think that would make a lot of sense.

oesteban commented on bids/layout/config/bids.json:
```diff
@@ -80,18 +80,18 @@
+        "sub-{subject}[/ses-{session}]/{datatype<dwi>|dwi}/sub-{subject}[_ses-{session}][_acq-{acquisition}]_{suffix<dwi>|dwi}.{extension<bval|bvec|json|nii.gz|nii>|nii.gz}",
+        "sub-{subject}[/ses-{session}]/{datatype<fmap>|fmap}/sub-{subject}[_ses-{session}][_acq-{acquisition}][_dir-{direction}][_run-{run}]_{fmap<phasediff|magnitude[12]|phase[12]|fieldmap>}.{extension<nii|nii.gz|json>|nii.gz}",
+        "sub-{subject}[/ses-{session}]/{datatype<fmap>|fmap}/sub-{subject}[_ses-{session}][_acq-{acquisition}][_ce-{ceagent}]_dir-{direction}[_run-{run}]_{fmap<epi>}.{extension<nii|nii.gz|json>|nii.gz}",
+        "sub-{subject}[/ses-{session}]/[{datatype<func|meg|beh>|func}/]sub-{subject}[_ses-{session}]_task-{task}[_acq-{acquisition}][_rec-{reconstruction}][_run-{run}][_echo-{echo}][_recording-{recording}]_{suffix<events>|events}.{extension<tsv|json>|tsv}",
```
Actually, I didn't add defaults to all suffixes - what I did is set the default value on the three suffixes that were only allowed to take one value. I agree this might turn out an unnatural behavior. I did it just for consistency with those ``{datatype<anat>|anat}`` which before were hard-coded and made `build_path` fail in strict mode when providing the `{'datatype': 'anat'}` since no replacement was found.

Do we still want to roll that back?

effigies commented:
I'm inclined to remove defaults from all suffixes. They're so central to the intended meaning of a file that imputing them seems unwise.

effigies commented on bids/layout/tests/test_path_building.py:
```diff
@@ -22,12 +22,12 @@ def test_invalid_file_construction(layout):
     ents = dict(subject='01', fakekey='foobar')
     with pytest.raises(ValueError):
-        layout.build_path(ents)
+        layout.build_path(ents, strict=True)
```
Could you add:

```Python
@pytest.mark.parametrize("strict", [True, False])
@pytest.mark.parametrize("validate", [True, False])
def test_insufficient_entities(layout, strict, validate):
    with pytest.raises(ValueError):
        layout.build_path({'subject': '01'}, strict=strict, validate=validate)
```

oesteban replied:
This might fall a bit outside the scope in the sense that the test is over `BIDSLayout.build_path` not over the actual function being modified. My worry is that, since I haven't been looking into `BIDSLayout.build_path` in-depth, the ValueError might be raised for other reasons too. Do you mind if I split this suggestion to its own PR?

(sorry, your review and my last comment happened simultaneously)

effigies replied:
That's fine.

oesteban commented:
@effigies, removed the defaults of suffixes (949fd1e5c1b15f5fa05908dea8d914bc0d2b23c5) and added one test-case&fix for a bug @garikoitz found yesterday (c25fb839f77b695952d74256f91d242043be6f25)

@tyarkoni - improved the docstring of the entities parameter (I realized I had missed out your suggestion, ac4ace8a21f44995d1f3af78cc02c81f3264e8bd)

Both, I added a warning for patterns like `{session<A|B|C>|D}` where the default value is not one of the valid values (f52189cc0183f285ed8defecc895e3bf3e73f774).

I think this is ready for merge. We can always go back to issues and add more tests within the continuation of this PR (#576).

oesteban commented:
Added the missing test on a new PR. Any objections?
